### PR TITLE
29 correct choropleth mapbox layer implementation

### PR DESCRIPTION
**Pull request by panosatha**
Tasks:

- [x] Make choropleth efficient with loading of zoom in area only and read of shapefiles with pyogrio
- [x] Add legend for choropleths
- [ ] Fix issue of not being able to hide the legend

### CODE REVIEW

**Comment by panosatha** [comment]
@maartenvanormondt can you check what happens with the legend of the choropleth? I followed your approach form the image layer but it doesn't seem to work. 

Another issue is a bug where while the legend is hidden, it reappears if you move the map.

File: src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py [hunk 1]
```diff
@@ -45,93 +47,123 @@ def make_overlay(self):
...
+        )
+        imgs.append(tf.shade(agg, cmap="red"))
         img = tf.stack(*imgs)
 
         path = os.path.dirname(file_name)
         name = os.path.basename(file_name)
         name = os.path.splitext(name)[0]
         export_image(img, name, export_path=path)
 
+        print(f"{time.time() - t0} to make overlay")
```
Comment (dladrichem) [inline src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py:83]: Do you still need this?

File: src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py [hunk 2]
```diff
@@ -147,30 +179,33 @@ def update(self):
...
 
+        t0 = time.time()
         if type(data) == str:
             # Read geodataframe from shape file
             self.gdf = read_dataframe(data)
-        else:    
+        else:
             # data has to be geodataframe
             self.gdf = data
+        print(f"{time.time() - t0} to load the dataframe")
```
Comment (dladrichem) [inline src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py:203]: Also this and probably all t0's are for manual testing I assume? These can be removed now then

File: src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py [hunk 3]
```diff
@@ -147,30 +179,33 @@ def update(self):
...
 
+        t0 = time.time()
         # Convert to spatialpandas geodataframe in wweb mercator
         self.data = sp.GeoDataFrame(self.gdf.set_crs(4326).to_crs(3857))
+        print(f"{time.time() - t0} to convert the dataframe")
```
Comment (dladrichem) [inline src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py:208]: Same

File: src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py [hunk 4]
```diff
@@ -12,38 +13,89 @@ def set_data(
...
-                self.line_color,
-                self.line_width,
-                self.line_opacity,
-                self.fill_opacity,
-                scaler
-            ],
-        )
+        # Prepare colormap
+
+        if not self.big_data:
```
Comment (dladrichem) [inline src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py:44]: Wouldn't it make more sense to put this in the update function as well?

File: src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py [hunk 5]
```diff
@@ -12,38 +13,89 @@ def set_data(
...
+                    self.line_color,
+                    self.line_width,
+                    self.line_opacity,
+                    self.fill_opacity,
+                    self.scaler,
+                    self.legend_title,
+                    self.unit,
+                ],
+            )
+        self.visible = True
```
Comment (dladrichem) [inline src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py:64]: Why do you first set self.visible, and then check for the same variable?

File: src/guitares/pyqt5/mapbox/geojson_layer_heatmap.py [hunk 6]
```diff
@@ -15,22 +16,24 @@ def set_data(self, data, density_property=""):
...
             arglist=[
                 self.map_id,
-                data,
+                self.data,
                 self.max_zoom,
-                density_property,
+                self.density_property,
             ],
         )
+        print(f"{time.time() - t0} to parse heatmap geojson to mapbox")
```
Comment (dladrichem) [inline src/guitares/pyqt5/mapbox/geojson_layer_heatmap.py:36]: Same

**Review by dladrichem — CHANGES_REQUESTED** [review]

File: src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py (hunk 1 above)
Comment (panosatha) [reply src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py:83]: now deleted

File: src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py (hunk 2 above)
Comment (panosatha) [reply src/guitares/pyqt5/mapbox/datashader_choropleth_layer.py:203]: correct

File: src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py (hunk 4 above)
Comment (panosatha) [reply src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py:44]: the reason I did this is that if the data are not big_data I don't want to update them. Only when there are big data I want to load only at the zoom level.

File: src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py (hunk 5 above)
Comment (panosatha) [reply src/guitares/pyqt5/mapbox/geojson_layer_choropleth.py:64]: that variable can be changed based on the GUI callbacks. Here by default I want to see the data when I set them to the layer. 

The update method though i called everytime the user changes the zoom level or pans the map. Then if the layer is not visible I don't want anything to happen.

Maybe not the most efficient way. But I had to work with how Maarten has set-up the whole Guitares package.

**Review by dladrichem — APPROVED** [review]

**Pull request merged**